lib/rucio/core/account.py

```diff
@@ -244,11 +245,11 @@ def list_identities(account, *, session: "Session"):
     ).where(
         models.IdentityAccountAssociation.account == account
     )
-    return [row._asdict() for row in session.execute(query)]
+    return [cast(IdentityDict, row._asdict()) for row in session.execute(query)]
```

Comment:
This is what occurs when I don't use `cast`:

```
Found 5 new problems in /home/runner/work/rucio/rucio/lib/rucio/core/account.py
  - 1 errors with message """Expression of type "list[Dict[str, Any]]" cannot be assigned to return type "list[IdentityDict]"""".
    Candidates: line 248
  - 1 errors with message """Expression of type "list[Dict[str, Any]]" cannot be assigned to return type "list[AccountAttributesDict]"""".
    Candidates: line 278
  - 1 errors with message """Expression of type "Dict[str, Any]" cannot be assigned to return type "UsageDict"
                               "Dict[str, Any]" is incompatible with "UsageDict"""".
    Candidates: line 377
  - 1 errors with message """Expression of type "list[dict[str, Any]]" cannot be assigned to return type "list[AccountUsageModelDict]"""".
    Candidates: line 398
  - 1 errors with message """Expression of type "list[Dict[str, Any]]" cannot be assigned to return type "list[UsageDict]"""".
    Candidates: line 424
```

See here for reference: https://github.com/rdimaio/rucio/actions/runs/8206991793/job/22447307300#step:9:10

Reply:
I think the `Dict[str, Any]` hint must be coming from [SQLAlchemy itself](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row._asdict). But by using `cast()` we effectively lose type checking:

```console
$ cat --number test.py
     1	#!/usr/bin/env python3
     2	
     3	from typing import TypedDict, cast
     4	
     5	class Struct(TypedDict):
     6	    foo: str
     7	    bar: int
     8	
     9	def test1() -> Struct:
    10	    return {'foo': 'foo', 'bar': 0}
    11	
    12	def test2() -> Struct:
    13	    return {'baz': None}
    14	
    15	def test3() -> Struct:
    16	    return cast(Struct, {'baz': None})
$ mypy test.py 
test.py:13: error: Missing keys ("foo", "bar") for TypedDict "Struct"  [typeddict-item]
test.py:13: error: Extra key "baz" for TypedDict "Struct"  [typeddict-unknown-key]
Found 2 errors in 1 file (checked 1 source file)
```

Does that mean that we need to avoid the use of `_asdict()` and `to_dict()` altogether? How annoying.

Reply:
Instead of using `cast`, this seems to work fine:
```
UsageDict(**session.execute(query).one()._asdict())
```

see test commit for reference: https://github.com/rdimaio/rucio/commit/cff7fa20131248acc2c4a06538c20b024d3c0b98

And pyright report: https://github.com/rdimaio/rucio/actions/runs/8233368875/job/22512678528

Reply:
I’m afraid that I failed to verify that it will raise an error in case of a mismatch:

```console
$ cat --number test.py 
     1	#!/usr/bin/env python3
     2	
     3	from typing import TypedDict
     4	
     5	from sqlalchemy import select
     6	from sqlalchemy.orm import Session
     7	
     8	from rucio.db.sqla.models import Identity
     9	from rucio.db.sqla.session import read_session
    10	
    11	
    12	# This is wrong.
    13	class IdentityDict(TypedDict):
    14	    foo: str
    15	    bar: int
    16	
    17	
    18	@read_session
    19	def list_identities(*, session: Session) -> list[IdentityDict]:
    20	    query = select(
    21	        Identity.identity
    22	    )
    23	    return [IdentityDict(**result._asdict()) for result in session.execute(query)]
    24	
    25	l = list_identities()
$ pyright test.py 
0 errors, 0 warnings, 0 informations
```

Reply:
This is odd, I found some sources online that seem to indicate that it should work:
- https://stackoverflow.com/questions/75254643/why-are-typeddict-types-without-field-and-with-notrequired-field-incompatible
- https://stackoverflow.com/questions/74495944/no-type-hint-with-return-values-for-typeddict-in-pycharm

And using extra keys in a TypedDict object construction (in the example, `**result._asdict()` are all extra keys to `IdentityDict`) should result in an error, according to PEP-589: https://stackoverflow.com/questions/71814658/python-typing-does-typeddict-allow-additional-extra-keys - the answer also mentions that pyright implements this.

I tried using a `dataclass` instead of a `TypedDict`, and I also don't get any errors raised from the incompatibility:

```
❯ cat --number test.py  
     1  #!/usr/bin/env python3
     2  from sqlalchemy import select
     3  from sqlalchemy.orm import Session
     4
     5  from rucio.db.sqla.models import Identity
     6  from rucio.db.sqla.session import read_session
     7  from dataclasses import dataclass
     8
     9  @dataclass
    10  class IdentityStruct:
    11      first: str
    12      second: int
    13
    14  @read_session
    15  def list_identities(*, session: Session) -> list[IdentityStruct]:
    16      query = select(
    17          Identity.identity
    18      )
    19      return [IdentityStruct(**result._asdict()) for result in session.execute(query)]
    20
    21  l = list_identities()                                                                                                                      


❯ python3 -m pyright test.py 
0 errors, 0 warnings, 0 informations 
```

Reply:
Some tests without the `sqla` components, just to check.

# 1. Incompatible TypedDict, constructor

```
#!/usr/bin/env python3
from typing import Any, TypedDict

class IncompatibleDict(TypedDict):
    foo: str
    bar: int

def print_dict(in_dict: dict[str, Any]) -> IncompatibleDict:
    d = IncompatibleDict(**in_dict)
    return(d)

test_dict = {
    "name": "test",
    "version": 3
}

l = print_dict(test_dict)
```

## Pyright doesn't complain
```
❯ python3 -m pyright test.py                                                                                                                      ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy complains that the dict type is unsupported for ** expansion
```
❯ python3 -m mypy test.py                                                                                                                         ─╯
test.py:9: error: Unsupported type "dict[str, Any]" for ** expansion in TypedDict  [typeddict-item]
Found 1 error in 1 file (checked 1 source file)
```

However, if I change the type annotation of `in_dict` from `dict[str, Any]` to `dict`, it behaves more as expected:
```
❯ python3 -m mypy test.py                                                                                                                         ─╯
test.py:9: error: Missing keys ("foo", "bar") for TypedDict "IncompatibleDict"  [typeddict-item]
Found 1 error in 1 file (checked 1 source file)
```

# 2. Incompatible TypedDict, cast
```
#!/usr/bin/env python3
from typing import Any, cast, TypedDict

class IncompatibleDict(TypedDict):
    foo: int
    bar: int

def print_dict(in_dict: dict[str, Any]) -> IncompatibleDict:
    d = cast(IncompatibleDict, in_dict)
    return(d)

test_dict = {
    "name": "test",
    "version": 3
}

l = print_dict(test_dict)
```

## Pyright doesn't complain
```
❯ python3 -m pyright test.py                                                                                                                      ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy doesn't complain
```
❯ python3 -m mypy test.py                                                                                                                         ─╯
Success: no issues found in 1 source file
```

# 3. Incompatible dataclass
```
#!/usr/bin/env python3
from typing import Any, cast
from dataclasses import dataclass

@dataclass
class IncompatibleClass:
    foo: int
    bar: int

def print_class(in_dict: dict[str, Any]) -> IncompatibleClass:
    d = cast(IncompatibleClass, in_dict)
    return(d)

test_dict = {
    "name": "test",
    "version": 3
}

l = print_class(test_dict)
```

## Pyright doesn't complain
```
❯ python3 -m pyright test.py                                                                                                                      ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy doesn't complain
```
❯ python3 -m mypy test.py                                                                                                                         ─╯
Success: no issues found in 1 source file
```

# Versions
```
❯ python3 -m pyright --version                                                                                                                    ─╯
pyright 1.1.353
```

```
❯ python3 -m mypy --version                                                                                                                       ─╯
mypy 1.9.0 (compiled: yes)
```

Reply:
Going back to static type checking - for test 4 (Incompatible model, runtime Pydantic test, works as expected), mypy and pyright do not detect issues by default:

```
❯ python3 -m mypy test.py                                                                                                                         ─╯
Success: no issues found in 1 source file
```

```
❯ python3 -m pyright test.py                                                                                                                      ─╯
0 errors, 0 warnings, 0 informations 
```

Reply:
# 6. Incompatible TypedDict, passing parameters directly, no extra parameters
```
#!/usr/bin/env python3
from typing import Any, TypedDict

class IncompatibleDict(TypedDict):
    foo: str
    bar: int

def print_dict() -> IncompatibleDict:
    test_dict = {
        "name": "test",
        "version": 3
    }
    d = IncompatibleDict(foo=test_dict['name'], bar=test_dict["version"])
    return(d)


l = print_dict()
```

## Pyright doesn't complain
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:13: error: Incompatible types (expression has type "object", TypedDict item "foo" has type "str")  [typeddict-item]
lib/rucio/core/test.py:13: error: Incompatible types (expression has type "object", TypedDict item "bar" has type "int")  [typeddict-item]
Found 2 errors in 1 file (checked 1 source file)
```

# 7. Incompatible TypedDict, passing parameters directly, extra parameters
```
#!/usr/bin/env python3
from typing import Any, TypedDict

class IncompatibleDict(TypedDict):
    foo: str
    bar: int

def print_dict() -> IncompatibleDict:
    test_dict = {
        "name": "test",
        "version": 3
    }
    d = IncompatibleDict(foo=test_dict['name'], bar=test_dict["version"], extra="unexpected")
    return(d)


l = print_dict()
```

## Pyright complains
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
/home/rdm/tech/rucio/rucio/lib/rucio/core/test.py
  /home/rdm/tech/rucio/rucio/lib/rucio/core/test.py:13:9 - error: No overloads for "__init__" match the provided arguments
    Argument types: (Unknown, Unknown, Literal['unexpected']) (reportCallIssue)
1 error, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:13: error: Extra key "extra" for TypedDict "IncompatibleDict"  [typeddict-unknown-key]
lib/rucio/core/test.py:13: error: Incompatible types (expression has type "object", TypedDict item "foo" has type "str")  [typeddict-item]
lib/rucio/core/test.py:13: error: Incompatible types (expression has type "object", TypedDict item "bar" has type "int")  [typeddict-item]
Found 3 errors in 1 file (checked 1 source file)
```

Reply:
# 8. Incompatible Pydantic model, passing parameters directly, static type checking
```
#!/usr/bin/env python3
from pydantic import BaseModel, ConfigDict

class IncompatibleModel(BaseModel):
    foo: int
    bar: int

def print_class() -> IncompatibleModel:
    test_dict = {
        "name": "test",
        "version": 3
    }
    return(IncompatibleModel(foo=test_dict['name'], bar=test_dict['version']))

l = print_class()
```

Pydantic complains here, as we saw earlier. From the point of view of the static type checkers:

## Pyright doesn't complain
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:13: error: Argument "foo" to "IncompatibleModel" has incompatible type "object"; expected "int"  [arg-type]
lib/rucio/core/test.py:13: error: Argument "bar" to "IncompatibleModel" has incompatible type "object"; expected "int"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

Reply:
# 9. Incompatible dataclass, passing parameters directly
```
#!/usr/bin/env python3
from typing import Any, cast
from dataclasses import dataclass, field

@dataclass
class IncompatibleClass:
    foo: int
    bar: int

def print_class() -> IncompatibleClass:
    test_dict = {
        "name": "test",
        "version": 3
    }
    d = IncompatibleClass(foo=test_dict['name'], bar=test_dict['version'])
    return(d)


l = print_class()
```

## Pyright doesn't complain
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy complains (ambiguously)
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:15: error: Argument "foo" to "IncompatibleClass" has incompatible type "object"; expected "int"  [arg-type]
lib/rucio/core/test.py:15: error: Argument "bar" to "IncompatibleClass" has incompatible type "object"; expected "int"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

Mypy complains even if I change `foo` to `str` - its problem seems to be that the type for `foo` and `bar` is not explicitly mentioned.

# 10. Incompatible dataclass, passing parameters directly
```
#!/usr/bin/env python3
from typing import Any, cast
from dataclasses import dataclass, field

@dataclass
class IncompatibleClass:
    foo: int
    bar: int

def print_class() -> IncompatibleClass:
    name = "test"
    version = 3
    d = IncompatibleClass(foo=name, bar=version)
    return(d)

l = print_class()
```

## Pyright complains
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
/home/rdm/tech/rucio/rucio/lib/rucio/core/test.py
  /home/rdm/tech/rucio/rucio/lib/rucio/core/test.py:13:31 - error: Argument of type "Literal['test']" cannot be assigned to parameter "foo" of type "int" in function "__init__"
    "Literal['test']" is incompatible with "int" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:13: error: Argument "foo" to "IncompatibleClass" has incompatible type "str"; expected "int"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

# 11. Incompatible TypedDict, passing parameters directly, adding explicitly types for the input arguments
```
#!/usr/bin/env python3
from typing import Any, TypedDict

class IncompatibleDict(TypedDict):
    foo: int
    bar: int

def print_dict() -> IncompatibleDict:
    name: str = "test"
    version: int = 3
    d = IncompatibleDict(foo=name, bar=version)
    return(d)


l = print_dict()
```

## Pyright complains
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
/home/rdm/tech/rucio/rucio/lib/rucio/core/test.py
  /home/rdm/tech/rucio/rucio/lib/rucio/core/test.py:11:30 - error: Argument of type "Literal['test']" cannot be assigned to parameter "foo" of type "int" in function "__init__"
    "Literal['test']" is incompatible with "int" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:11: error: Incompatible types (expression has type "str", TypedDict item "foo" has type "int")  [typeddict-item]
Found 1 error in 1 file (checked 1 source file)
```

Reply:
# Conclusion
- Static type checking: mypy seems to be slightly stricter compared to pyright - I think it would be worth considering we move to mypy.
- Runtime type checking: I think this should be necessary to catch cases where static type checking would not be enough. I opened an issue to discuss this: https://github.com/rucio/rucio/issues/6544

In the long term, I think the best approach would be to use mypy+pydantic and use pydantic models for struct typing. For now, it seems to be fine to use `TypedDict`, passing arguments directly, instead of passing them as kwargs, and adding explicit type annotations for the input arguments, so that the static type checkers are able to notice any issues.

Reply:
Relevant thread https://github.com/rucio/rucio/discussions/6615